### ISSUE TITLE
What comes first: the SRS or the SystemInformation?

### ISSUE THREAD

**Issue by JacquesCarette** [JacquesCarette/Drasil#1086]
While investigating the duplication of information in `SystemInformation`, I can across the following interesting problem. I will use GamePhysics as an example, but it is representative of the others as well.

Notice that the definition of the SRS (as a `DocDesc`) has the following line
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L104

which uses `dataDefns`, which is itself a list of `DataDefinition` hand-crafted in `GamePhysics.DataDefs`
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L66

This `dataDefns` (essentially) is re-used, indirectly, when defining `game_datadefn`
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L135-L136

This is mostly good, in that it ensures consistency: `game_datadefn` will only contain those `DataDefinition` that actually appear in the SRS.

Futher down, the `SystemInformation` is assembled. There are actually 3 fields in it that contain a list of `DataDefinition`s (which is clearly too many...), initialized this way:
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L174
and also
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L181-L182
https://github.com/JacquesCarette/Drasil/blob/389723cb501bddb0c97a915c571b6cac01810b07/code/drasil-example/Drasil/GamePhysics/Body.hs#L197-L209

The part to notice is that the `_datadefs` field uses `dataDefns` directly and the two `ChunkDB` instead use `game_defn`. These happen to be consistent.  And the `everything` happens to all be used (wrt data definitions), so there's no difference (unlike for other fields, but that's a different issue).

The issue here is: rather than testing for consistency post-facto, why risk a divergence? But that brings about the issue in the title: what is to be taken as the main source of 'truth'?  The SRS or the SystemInformation?

Answering this, I think, requires thinking about the (idealized?) process of 'programming' in Drasil. Here is an informal, current, view:
1. create/use some general information about some domains (in `drasil-data` in `Data.Drasil.*` files)
2. create some specific information about a domain (`Data.Drasil.GamePhysics.Concept`, `.DataDefs`, etc)
3. a) assemble lots of this information into an `SRS`
    b) assemble all of this information as well as some design decisions, into a `SystemDescription`
4. actually make the SRS document (uses both the SRS data and the SystemDescription)
5. actually make the code (see GlassBR) which only uses the SystemDescription (!)

This doesn't seem like the right process. Something is wrong - but there are many possible ways of 'fixing' that. And that fix isn't just a code thing, as has deep implications on how Drasil ought to be used to go from knowledge to SE artifacts.

**Comment by smiths** [comment]
@JacquesCarette, I haven't forgotten this issue; it just slipped down in my "to do" list.  I don't understand all of the details, but I can see that what we currently have is not the idealized process of programming in Drasil.

We have been using the SystemInformation for additional tasks over this summer.  In particular, we are attempting to generate the requirements from the system information.  This seems like a step in the right direction to me.

The SRS is just one view of the knowledge.  I think it has taken too large of a role in Drasil, mostly because it was the most complete part in the "manual" case studies, and because we did it first.  I think the knowledge is what should come first.  It seems like we are putting our knowledge in System Information.

Is the following too simple a process?

1. Assemble all information into a `SystemDescription`
2. Make the SRS document
3. Make the Code

We could have other steps for other artifacts (like design documentation, test cases, etc.).

We should probably have a meeting to discuss this further.  We are meeting on Thursday morning (Aug 15) already, so maybe we could add a bit of discussion on this topic to the agenda?

**Comment by JacquesCarette** [comment]
Yes, it is too simple. I think we will need a multi-pass process that first assembles a particular "kind" of information, then transforms it (with a recipe), then adds more information, transform, add, etc. Interspersed in there will be the generation of various artifacts; they can be generated as soon as all the information they rely on has been gathered.

This process is also more modular, as it allows earlier "information bundles" to be re-used in later changes when changes (like different low-level design choices) are made that do not change anything previous in the 'pipeline'.

**Comment by cd155** [comment]
To Record what we discussed on Dec 13th, 2021 group meeting about this issue.

# Create ideal workflow
### Background
Drasil was first created to produce SRS, and then start to generate code. Then, it be could become a software product, which can belong to a tree of Drasil family product/system. In this tree of Drasil family product, is there a single data structure that could contain all information for all products?

We may think SRS might give enough information, but the requirement only describe what it wants without giving how it should be done, so it doesn't have enough information for this single data structure. Currently, the `System Information` serves as this single data structure that contains all information. 

### Current Situation 
The giant `system information` collects a multitude of pieces of information, whenever we need it, we extract the information from the `system information`. The choices data structure is separated from `system information`.

### Unsolved Questions
1. Is the current workflow of `system information` right workflow?
2. What information should include in the `system information`?
3. What information should include in the requirements (e.g. theory manual, how to generate make files)?
